### DISCUSSION TITLE
Add npm init option to generate a .gitignore file containing node_modules

### DISCUSSION THREAD

**Discussion by sdavids** [npm/rfcs#742] [RRFC]
Add an npm init option to generate a _.gitignore_ file containing `node_modules`.

Either via a CLI option or an npm config flag.

Something along the lines of `npm init --generate-gitignore`

The generated (or updated) _.gitignore_ file would have an entry:

```
node_modules/
```

This could even be the default going forward.

**Comment by ljharb** [comment]
The default gitignore file that github generates has node_modules in it already (assuming you tell it "node" when you create a new repo). It seems strange for `npm init` to do git stuff.

**Reply by sdavids** [reply]
Not everyone uses GitHub…

> (assuming you tell it "node" when you create a new repo)

```shell
$ gh repo create test --private -g node
HTTP 422: Repository creation failed. (https://api.github.com/user/repos)
gitignore_template is an unknown gitignore template.
```

Well, now one has to look into the doc again, after remembering to set the flag in the first place.

https://cli.github.com/manual/gh_repo_create

Doesn't tell the available templates … so further down the rabbit hole…

https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files

> GitHub maintains an official list of recommended .gitignore files for many popular operating systems, environments, and languages in the "github/gitignore" public repository.

https://github.com/github/gitignore/blob/main/Node.gitignore

⇓

```shell
$ gh repo create test --private -g Node
```
Now one has a `.gitignore` file containing 130 lines—yay! 😉

---

If you want to be fancy, `npm init` could execute `git rev-parse --is-inside-work-tree` and if `true` then create the `.gitignore` with `node_modules/`

---

Yes, not everyone uses git. And yes, why should one tool create configuration for another tool.

But I would wager the absolute majority of npm projects are versioned via git.

**Reply by ljharb** [reply]
Yes, those 130 lines are ideal.

**Comment by wesleytodd** [comment]
I am pretty sure the roadmap here is to move project initialization beyond the most bare bones `package.json` out of npm cli. A few folks have made very slow progress on this plan here: https://github.com/pkgjs/create-pkg/issues/7

At one point there was a pretty general agreement that someone other than the npm cli team should own the setup and let the npm team focus on more important aspects of what npm does. I believe that goal is still well shared with the npm team (I just don't want to directly speak for them).

As part of that work [I created this](https://www.npmjs.com/package/create-git) which might suite your needs a bit. In the long run our goal is for you to be able to run `npm init pkg` and get all the stuff from that diagram in the issue posted above.

So all this is to say, I don't think npm adding this to the base `npm init` is the right technical direction for them to take.

**Reply by wesleytodd** [reply]
I looked at that repo and realized it needed some love, so just pushed the 1.0.0 branch with mostly updated things. Ideally I will get around to automating the upkeep of this soon, I just started with the `create-package-json` package first since that is seeing some good active work with the package metadata collab space.
